### ISSUE TITLE
self signed certificate error

### ISSUE THREAD

**Issue by eric-elsholz** [brianc/node-postgres#2009]
Hi. Sorry ahead of time, I don't have a lot of detail here. We updated our packages today and afterwards started seeing a connection error from Sequelize ORM: "self signed certificate". We do in fact use a self signed certificate but this has never been a problem for us in the past. We downgraded from pg@7.13.0 to pg@7.4.3 and then we were able to connect again.

**Comment by ianwremmel** [comment]
If I’m reading https://github.com/brianc/node-postgres/commit/bf029c827049ca16add0a862d40f4e60dfd9e602 correctly, I think there may have been a bunch of options (including rejectUnauthorized) being inadvertently set to falsy values. They are now not set at all unless you pass them in. The last release was a breaking change, but it’s probably easiest to fix by passing rejectUnauthorized: false to pg.

**Comment by ianwremmel** [comment]
Sorry, I think I actually meant checkServerIdentity.

**Comment by charmander** [comment]
> I think there may have been a bunch of options (including rejectUnauthorized) being inadvertently set to falsy values

`tls.connect` treats `rejectUnauthorized: undefined` as `rejectUnauthorized: false`? … It does. Oh no. :(

Well, it’s a good thing that was fixed. We should probably add a test for it. It might also be worth a security announcement (@brianc).

**Comment by powerc9000** [comment]
God I thought I was going crazy! Thanks for this info rolling back to 7.12 worked for me.

**Comment by eldimious** [comment]
I am facing the same issue. Rolling back to `7.12.1` worked for me too.

**Comment by znarf** [comment]
Spotted that in our staging system (heroku), happy to give more details.

**Comment by BrandonZacharie** [comment]
I'm using Sequelize as well and I got the "self signed certificate in chain" error until I rolled back to 7.12.1

**Comment by Kinetic-Pulse** [comment]
This one one of the most painful bugs I had to identify.
Probably shouldnt do too many updates at once but ye....
Downgrading is the way for now ~7.12.1

**Comment by ptpt** [comment]
Same here. Very hard to identify. The only error message we got is:
```
Error: self signed certificate in certificate chain
    at TLSSocket.onConnectSecure (_tls_wrap.js:1321:34)
    at TLSSocket.emit (events.js:210:5)
    at TLSSocket.EventEmitter.emit (domain.js:476:20)
    at TLSSocket._finishInit (_tls_wrap.js:794:8)
    at TLSWrap.ssl.onhandshakedone (_tls_wrap.js:608:12) {
  code: 'SELF_SIGNED_CERT_IN_CHAIN'
}
```
And we noticed that it occurs when we query PSQL.

Downgraded to `pg@7.4.3` and it works now.

**Comment by jsanta** [comment]
If it may help someone else, we just  added the `rejectUnauthorized: false`, to our sequelize connection configuration:
```
sequelize: {
    uri:
      'postgres://postgres_user:postgres_user_password@postgres_server:5432/postgres_database',
    options: {
      database: 'postgres_database',
      dialect: 'postgres',
      host: 'postgres_server',
      port: 5432,
      username: 'your_postgres_user',
      password: 'your_postgres_user_password',
      pool: {
        max: 3,
        min: 1,
        idle: 10000,
      },
      dialectOptions: {
        ssl: {
          require: true,
          // Ref.: https://github.com/brianc/node-postgres/issues/2009
          rejectUnauthorized: false,
        },
        keepAlive: true,        
      },      
      ssl: true,
      define: {
        timestamps: false,
      },
    },
  },
```
lets say the error is still there, but we are not being notified, and the backend is not crashing anymore. Just adjust the configuration according to your own needs.
Best regards from Chile.

**Comment by brianc** [comment]
Gah, I apologize for this.  I'm going to roll out a backwards-compatible (to the 7.x line) semver patch on top of this one to make sure it behaves exactly as it should.

@charmander what do you mean about security announcement?  This makes things __more__ secure now right? By forcing unauthorized cert rejection, where as before it was ignoring it?  We should probably make the defaults be the same as the node.js defaults in v8.0.

**Comment by brianc** [comment]
Okay - released `pg@7.14.0` which reverts `pg@7.13.0` 🤕 sorry about that!

I'd like to get self-signed and "proper" ssl connections tested in travis.  Anyone have a docker file handy that sets both of those things up?  Then we can be covered on this behavior going forward.

**Comment by charmander** [comment]
> @charmander what do you mean about security announcement? This makes things **more** secure now right? By forcing unauthorized cert rejection, where as before it was ignoring it?

@brianc Yep, it does! I meant an announcement that pg 0.8.7 through 7.12.0 don’t check certificates.

**Comment by sehrope** [comment]
+1 to having the default in a new major being to validate so that it's an explicit action to disable things.

FYI, the default for libpq (ex: via psql) is not to validate either so the default behavior of this module till now is not totally crazy either. May be worth calling that out in any security announcement: https://www.postgresql.org/docs/current/libpq-ssl.html (Still a +1 to making a security note calling out the behavior)

**Comment by hghammoud** [comment]
any news on this subject? any pg version that properly supports self signed certs?

**Comment by brianc** [comment]
I'll add this to pg@8.0 which is under active development & should be released in a week or two, so will look at this shortly sorry for the delay

**Comment by charmander** [comment]
> any news on this subject? any pg version that properly supports self signed certs?

@hghammoud What’s “properly”? If you mean the behaviour of not verifying certificates, that was [already restored in 7.14.0](https://github.com/brianc/node-postgres/issues/2009#issuecomment-556080489). If you want verification, you’ll have to specify options like:

```js
{
    ssl: {
        rejectUnauthorized: true,
        ca: /* the certificate */,
    },
}
```

and that’s not going to change with 8.0 except for `rejectUnauthorized: true` being implied (maybe). It’s already done “properly” with respect to the difference between self-signed and trusted by the system.

**Comment by brianc** [comment]
@charmander - I was looking at the postgres docs...

https://www.postgresql.org/docs/9.1/libpq-ssl.html

Particularly at "Table 31-1. SSL Mode Descriptions" - I think specifying `ssl: true` and having `rejectUnauthorized` to `false` is actually pretty much in line w/ how libpq does things.  I think what we should do is [here](https://github.com/brianc/node-postgres/blob/master/packages/pg/lib/connection-parameters.js#L36) is support being able to opt in to this behavior via the `PGSSLMODE=verify-ca"...at that point instead of returning `true` we should return `{ rejectUnauthorized: true }` to use it as the config.  

Other than that I'm not even totally sure deprecating the ability to leave it `undefined` it is really the right thing as it behaves the same in libpq....`undefined` implies `false`.  On one hand it's nice to let users know "hey, are you sure you don't want to reject unauthorized certificates?" but libpq doesn't do that either, it just happily goes along and doesn't verify unauthorized certs unless explicitly asked to do so.  as far as I can see this is the final kinda dangling issue WRT me releasing a new semver minor `7.x` w/ the warning included as well as doing (or not) doint any changes for 8.0.

**Comment by charmander** [comment]
It’s in line with how libpq does things, but I still think it’s better to be secure by default – in line with Node’s tls. I was surprised by it, at least, but maybe that was influenced by reading the code (which can make for misleading reading because the change to not verifying by default was an accident). Parsing libpq’s connection strings into the appropriate settings would be a good change, though, yeah. (And to be clear, the deprecating the ability to leave it undefined isn’t supposed to be permanent, right? It’ll default to `rejectUnauthorized: false` with the new major version.)

**Comment by brianc** [comment]
> (And to be clear, the deprecating the ability to leave it undefined isn’t supposed to be permanent, right? It’ll default to `rejectUnauthorized: false` with the new major version.)

Well that's why I brought it up - I'm not sure we _should_ default to `rejectUnauthorized: false` if you do something like

```
const client = new Client({ ssl: true })
```

I guess we could do that....you're right its more secure by default.  

I think in that case we should make sure if you use the environment variables:

```
PGSSLMODE=require 
PGSSLMODE=verify-ca
```

We pass the proper value of either ` { rejectUnauthorized: false }` or ` { rejectUnauthorized: true }` into the ssl config so _that_ stays in line with how the environment variable works.  So, yeah in 8.0 I'll need to get some better tests around ssl handshakes in their various forms and then fix the behavior of the environment vars to do what's "expected" in all cases and make sure native and pure js line up in behavior...that'll be the last thing before 8.0!  I'll start working on this this week...it's a bit more fiddly that the other changes because SSL is sorta complex by design so...it'll take a bit (but not too long!)

**Comment by brianc** [comment]
To elaborate on these changes a bit I think it should be like...

```
const client = new Client({ ssl: true })
// default ssl config = { rejectUnauthorized: true }
```

```
process.env.PGSSLMODE='require'
// default ssl config = { rejectUnauthorized: false }
```

```
process.env.PGSSLMODE = 'verify-ca'
// default ssl config = { rejectUnauthorized: true }
```

I think in every case we want the connection parameters (which parse env vars) to generate a _default_ ssl config, and in every case we want to allow any custom `ssl: { ... }` config passed to the client to overwrite the default config and extend it.

e.g.

```
PGSSLMODE=require
const client = new Client({
  ssl: {
     rejectUnauthorized: true
  }
})
```
should make it reject unauthorized.  Anything you pass explicitly to `{ ssl: { ... } }` takes precedence over environment variables and has the "final say" in the ssl config.

**Comment by brianc** [comment]
I'll document these permutations and also include a note to discourage the use of `new Client({ ssl: true })` as the meaning is ambiguous.  Perhaps we deprecate it in 8.0 in favor of environment variables or a more explicit full `ssl: { ... }` config

**Comment by jylopez** [comment]
In case it's not obvious (it wasn't to me), this is the solution:
```
const pool = new pg.Pool({
  connectionString: process.env.DATABASE_URL,
  ssl: {
    rejectUnauthorized: false
  }
});
```

Specifically, when trying to provision a heroku postgres server. Heroku's [doc](https://devcenter.heroku.com/articles/getting-started-with-nodejs#provision-a-database) is not updated. It **incorrectly** says to put:
```
const pool = new Pool({
  connectionString: process.env.DATABASE_URL,
  ssl: true
});
```

**Comment by fapdev77** [comment]
Also not obvious to me... Just a small correction use:

const pool = new Pool({ .... 

instead of

const pool = new pg.Pool({ ...

Worked to me.

Thanks @jylopez

**Comment by DiegoRBaquero** [comment]
This change should have been a semver major.

We use ssl in some envs only, and in those where we do we use self-signed. If we pass `ssl: {
    rejectUnauthorized: false
  }` we are enabling ssl, which won't work in non ssl envs. We use connection strings.

**Comment by charmander** [comment]
The original issue was fixed in pg 7.14.0, and the default change was made in 8.0. Support for the rest of the `sslmode` options is in #2105 (I think).

**Comment by johnschoeman** [comment]
> In case it's not obvious (it wasn't to me), this is the solution:
> 
> ```
> const pool = new pg.Pool({
>   connectionString: process.env.DATABASE_URL,
>   ssl: {
>     rejectUnauthorized: false
>   }
> });
> ```
> 
> Specifically, when trying to provision a heroku postgres server. Heroku's [doc](https://devcenter.heroku.com/articles/getting-started-with-nodejs#provision-a-database) is not updated. It **incorrectly** says to put:
> 
> ```
> const pool = new Pool({
>   connectionString: process.env.DATABASE_URL,
>   ssl: true
> });
> ```

For anyone coming here trying to fix a self-signed-certificate error on their Heroku + node + pg production instance, the recommended solution is to use `pg < 8.0.0.` and *not* turn off SSL altogether.

**Comment by charmander** [comment]
> the recommended solution is to use pg < 8.0.0

@johnschoeman Whose recommended solution? I would recommend what @jylopez wrote (`ssl: {rejectUnauthorized: false}`).

**Comment by johnschoeman** [comment]
@charmander , heroku's recommendation. Their docs say the preferred way is to use pg < 8.0.0: https://devcenter.heroku.com/articles/getting-started-with-nodejs#provision-a-database, apologies for not referencing that in the initial comment.

> I would recommend what @jylopez wrote (ssl: {rejectUnauthorized: false}) if you want the pg 7.x behaviour, not downgrading to pg 7.x.

Does this mean that these are equivalent? I may be confused with the API, but I read `ssl: { rejectUnauthorized: false }` as allowing requests without SSL certs.

**Comment by DiegoRBaquero** [comment]
They are not equivalent. Using `ssl: { rejectUnauthorized: false }` means enabling SSL but not rejecting self-signed.

We have different envs with different ssl/non-ssl settings, connection string is used, this broke things.

**Comment by brianc** [comment]
> heroku's recommendation. Their docs say the preferred way is to use pg < 8.0.0:

ah what the hell? I'll email them about this.  Not good to recommend an old version, particularly when it's less secure out of the box.

**Comment by brianc** [comment]
> They are not equivalent. Using `ssl: { rejectUnauthorized: false }` means enabling SSL but not rejecting self-signed.

Yeah that's correct - that's what you want to do with heroku.  What do you mean you have different envs with different ssl/non-ssl settings? Downgrading to `7.x` is going to cause you problems as more fixes & features continue to be added to `8.x` and beyond.  I've reached out to heroku about this.  I'm a bit annoyed they didn't mention anything here to me, but just updated their docs to have people install an old version....

What issues is it causing you exactly? I'd like to understand better so I can recommend a fix.

**Comment by brianc** [comment]
I sent in a support ticket on my paid heroku account, submitted feedback to the documentation article, and tweeted at them...hopefully I hear something back so I can help get this taken care of and clarity delivered to everyone.

**Comment by brianc** [comment]
> In case it's not obvious (it wasn't to me), this is the solution:

Sorry about that!  I tried to document [the change](https://node-postgres.com/announcements#2020-02-25) here.  I'm gonna try working with heroku to get their docs updated.

**Comment by jylopez** [comment]
@brianc Hey no problem man. You're doing a great job. Thanks for caring!

**Comment by DiegoRBaquero** [comment]
> > They are not equivalent. Using `ssl: { rejectUnauthorized: false }` means enabling SSL but not rejecting self-signed.
> 
> Yeah that's correct - that's what you want to do with heroku. This is exactly what the old behavior on `pg@7.x` did when you supplied ` { ssl: true }`. What do you mean you have different envs with different ssl/non-ssl settings? Downgrading to `7.x` is going to cause you problems as more fixes & features continue to be added to `8.x` and beyond. I've reached out to heroku about this. I'm a bit annoyed they didn't mention anything here to me, but just updated their docs to have people install an old version....
> 
> What issues is it causing you exactly? I'd like to understand better so I can recommend a fix.

Hey, thanks for answering

Right, so first, I'm not using Heroku at all. We are using Aptible, GCP and AWS. We have different settings in different envs, some support SSL (self signed) and other don't. We use connection strings in env variables. So there's no way to remain using that right now whille passing this setting. We have over 50 microservices in over 6 different env, so it's a lot of work to put this setting where we have SSL, ignore it where not.

We have massive -> pg-promise > node-pg. pg-promise updated driver to 8 in a minor version and screwed us (https://github.com/vitaly-t/pg-promise/issues/711). Now we are stuck in massive 6.2.0 because of this.

x2 Thank you for caring!

**Comment by brianc** [comment]
hmmmm is there an environment variable you think I could add to make this controllable too or easier for you to roll out?  I know psql supports [`PGSSLMODE`](https://www.postgresql.org/docs/9.0/libpq-connect.html#LIBPQ-CONNECT-SSLMODE) with a bunch of different options.  I could add something like `PGSSLMODE=no-verify` which would be the eqiv of passing `{ rejectUnauthorized: false }` in as the ssl param?  Then you could update env vars and not have to udpate code or do something like reach down into massive -> pg-promise -> node-pg to figure out how to thread the `ssl` config option down through all the layers.

They do have `PGSSLMODE=require` which _can_ fallback to not checking if there is no CA present.  I'm not sure how postgres checks for a root cert....I could likely dig in postgres client C code & find it...but w/ such limited time that's kinda lower on my priority list than making pg faster, which is what I have in the works coming up.

**Comment by johnschoeman** [comment]
> Sorry about that! I tried to document [the change](https://node-postgres.com/announcements#2020-02-25) here. I'm gonna try working with heroku to get their docs updated.

@brianc , thanks for the prompt response! This clears the confusion for me. 💯 I took a pass at updating the docs, if you're interested; I am new to node-postgres, and it wasn't clear to me to look in the announcements for this info.  https://github.com/brianc/node-postgres-docs/pull/71

**Comment by nabilfreeman** [comment]
Is there a way to trust some self-signed certificates and not others?

Forgive me if I don't understand SSL well enough, but the way I understand it is that my RDS instances from AWS have self signed certificate as part of their provisioning process and I've downloaded and trusted on my server. This is the only option for SSL certs on RDS.

So what I would expect is that those certificates don't throw an error, but if another certificate appeared with a different hash that one would be rejected as a security risk.

Or is it trivial to spoof self-signed certs?

**Comment by charmander** [comment]
@nabilfreeman Specify the certificate as the [`ca` option](https://nodejs.org/api/tls.html#tls_tls_createsecurecontext_options), like this:

```js
new pg.Pool({
  ssl: {
    ca: fs.readFileSync('cert.pem'),
  },
})
```

(this can also be an array)

**Comment by surferwat** [comment]
> In case it's not obvious (it wasn't to me), this is the solution:

After also adding `PGSSLMODE=no-verify` to `.env` worked for me.

**Comment by surferwat** [comment]
> In case it's not obvious (it wasn't to me), this is the solution:
> 
> ```
> const pool = new pg.Pool({
>   connectionString: process.env.DATABASE_URL,
>   ssl: {
>     rejectUnauthorized: false
>   }
> });
> ```
> 
> Specifically, when trying to provision a heroku postgres server. Heroku's [doc](https://devcenter.heroku.com/articles/getting-started-with-nodejs#provision-a-database) is not updated. It **incorrectly** says to put:
> 
> ```
> const pool = new Pool({
>   connectionString: process.env.DATABASE_URL,
>   ssl: true
> });
> ```

After also adding PGSSLMODE=no-verify to .env worked for me.

**Comment by brianc** [comment]
> After also adding `PGSSLMODE=no-verify` to `.env `worked for me.

👍 
Working w/ heroku on getting the docs updated.  I didn't ever intend to require code changes to heroku apps or configs when upgrading to pg@8.0....so trying to make it easier now by adding the `PGSSLMODE=no-verify`.  It can also be added to the connection string `?ssl=no-verify`

**Comment by Kellym-Kainos** [comment]
Does anyone know if it's possible to use nodes built in certs here:
```
  ssl: {
    rejectUnauthorized: true,
    ca: [path to ca]
  } ```

I've been able to use amazon rds cert but would be good to be able to use node certs so we dont have to update every 4/5 years

**Comment by brianc** [comment]
> Does anyone know if it's possible to use nodes built in certs here:

Howdy! You should be able to pass any of the config options that node's ssl constructor takes directly through - `pg` will just pass them in unaltered if you pass an object  as the `ssl` option to the pg client/pool constructor.

**Comment by Kellym-Kainos** [comment]
@brianc thanks however when i tried using one of these certs https://github.com/nodejs/node/blob/master/src/node_root_certs.h I'm still getting ```self signed certificate in chain``` error.

Do you ignore built in node certs?

**Comment by boromisp** [comment]
You can access the default root certificates as [`tls.rootCertificates`](https://nodejs.org/api/tls.html#tls_tls_rootcertificates) and you can extends the default root certificate list with the [`NODE_EXTRA_CA_CERTS=file`](https://nodejs.org/api/cli.html#cli_node_extra_ca_certs_file) environment variable.

I don't think the AWS RDS server certificates are signed with any generally trusted root certificates, so your only option is to download and explicitly trust one of the root certificates [provided by amazon](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/UsingWithRDS.SSL.html).

Based on [this serverfault answer](https://serverfault.com/a/814568) on you can view your database's certificate chain with this command:

```shell
echo "" | openssl s_client -starttls postgres -connect EXAMPLE.COM:5432 -showcerts
```

**Comment by valentincostam** [comment]
This worked for me:

```js
const isProduction = process.env.NODE_ENV === 'production'

const pool = new Pool({
  connectionString: process.env.DATABASE_URL,
  ssl: isProduction ? {rejectUnauthorized: false} : false
})
```

This way you prevent from getting the following error when your app is running on Heroku:

```
Error: self signed certificate
```

And also this error when it's running on localhost:

```
Error: The server does not support SSL connections
```

**Comment by koistya** [comment]
When using a self-signed certificate, don't forget to specify `servername`. For example, in the case with Google Cloud SQL it would be `{GOOGLE_GLOUD_SQL_PROJECT}:{GOOGLE_CLOUD_SQL_INSTANE}`:

```js
new pg.Pool({
  ssl: env.PGSSLMODE === "verify-ca" && {
    cert: fs.readFileSync(env.PGSSLCERT, "ascii"),
    key: fs.readFileSync(env.PGSSLKEY, "ascii"),
    ca: fs.readFileSync(env.PGSSLROOTCERT, "ascii"),
    servername: env.PGSERVERNAME, // e.g. "example:pg12"
  }
});
```

**Reference**: [Google Cloud SQL - Tips & Tricks](https://medium.com/@koistya/google-cloud-sql-tips-tricks-d0fe7106c68a)

**Comment by adrianhorning08** [comment]
I am using Digital Oceans App Platform and Managed Database using `"pg":` "^8.5.1",` and this works:
(specifying user, host, db, pw, and port
```
const pool = new Pool({
  user: process.env.DB_USER,
  host: process.env.DB_HOST,
  database: process.env.DB,
  password: process.env.DB_PASSWORD,
  port: process.env.DB_PORT,
  ssl: {
    rejectUnauthorized: true,
    ca: fs.readFileSync("ca_cert.crt").toString(),
  },
});
```

but this doesn't (just putting the connection string)
```
const pool = new Pool({
  connectionString: ***sslmode=require,
  ssl: {
    rejectUnauthorized: true,
    ca: fs.readFileSync("ca_cert.crt").toString(),
  },
});
```
So yeah that `sslmode=require` appending on the connection string I think is messing it up.

Why would the first one work and the second doesn't?

Wanted to put this here in case some poor soul (like future me) needs it.

**Comment by seamoss** [comment]
> I am using Digital Oceans App Platform and Managed Database using `"pg": "^8.5.1",` and this works:
> (specifying user, host, db, pw, and port)
> 
> ```
> const pool = new Pool({
>   user: process.env.DB_USER,
>   host: process.env.DB_HOST,
>   database: process.env.DB,
>   password: process.env.DB_PASSWORD,
>   port: process.env.DB_PORT,
>   ssl: {
>     rejectUnauthorized: true,
>     ca: fs.readFileSync("ca_cert.crt").toString(),
>   },
> });
> ```
> 
> but this doesn't (just putting the connection string)
> 
> ```
> const pool = new Pool({
>   connectionString: ***sslmode=require,
>   ssl: {
>     rejectUnauthorized: true,
>     ca: fs.readFileSync("ca_cert.crt").toString(),
>   },
> });
> ```
> 
> So yeah that `sslmode=require` appending on the connection string I think is messing it up.
> 
> Why would the first one work and the second doesn't?
> 
> Wanted to put this here in case some poor soul (like future me) needs it.

Also seeing this here.  Refactored to parse strings to dicts for now, but a strange behavior, indeed.

**Comment by Yadro** [comment]
`NODE_TLS_REJECT_UNAUTHORIZED=0` in process.env resolved this issue for me.

**Comment by PrimeTimeTran** [comment]
Same issue for me. My set up is different as I'm using it in a Node/TypeScript. Here are my Heroku logs.

- "sequelize": "6.3.5",
- "typescript": "4.0.3" 

```
2021-01-19T08:53:42.208018+00:00 heroku[router]: at=info method=OPTIONS path="/api/auth/sign-in" host=cs-lms.herokuapp.com request_id=121648d2-eb3f-47c9-81d1-bf1f3c85a532 fwd="14.169.111.180" dyno=web.1 connect=1ms service=12ms status=204 bytes=367 protocol=https
2021-01-19T08:53:42.571888+00:00 heroku[router]: at=info method=POST path="/api/auth/sign-in" host=cs-lms.herokuapp.com request_id=92e8a79e-62c9-467a-abfc-fc9205cacc1b fwd="14.169.111.180" dyno=web.1 connect=1ms service=101ms status=500 bytes=984 protocol=https
2021-01-19T08:53:42.567837+00:00 app[web.1]: ConnectionError [SequelizeConnectionError]: self signed certificate
2021-01-19T08:53:42.567852+00:00 app[web.1]: at Client._connectionCallback (/app/node_modules/sequelize/lib/dialects/postgres/connection-manager.js:184:24)
2021-01-19T08:53:42.567853+00:00 app[web.1]: at Client._handleErrorWhileConnecting (/app/node_modules/pg/lib/client.js:296:19)
2021-01-19T08:53:42.567854+00:00 app[web.1]: at Client._handleErrorEvent (/app/node_modules/pg/lib/client.js:306:19)
2021-01-19T08:53:42.567854+00:00 app[web.1]: at Connection.emit (events.js:314:20)
2021-01-19T08:53:42.567855+00:00 app[web.1]: at TLSSocket.reportStreamError (/app/node_modules/pg/lib/connection.js:53:12)
2021-01-19T08:53:42.567855+00:00 app[web.1]: at TLSSocket.emit (events.js:314:20)
2021-01-19T08:53:42.567856+00:00 app[web.1]: at emitErrorNT (internal/streams/destroy.js:92:8)
2021-01-19T08:53:42.567856+00:00 app[web.1]: at emitErrorAndCloseNT (internal/streams/destroy.js:60:3)
2021-01-19T08:53:42.567857+00:00 app[web.1]: at processTicksAndRejections (internal/process/task_queues.js:84:21) {
2021-01-19T08:53:42.567857+00:00 app[web.1]: parent: Error: self signed certificate
2021-01-19T08:53:42.567858+00:00 app[web.1]: at TLSSocket.onConnectSecure (_tls_wrap.js:1502:34)
2021-01-19T08:53:42.567858+00:00 app[web.1]: at TLSSocket.emit (events.js:314:20)
2021-01-19T08:53:42.567859+00:00 app[web.1]: at TLSSocket._finishInit (_tls_wrap.js:937:8)
2021-01-19T08:53:42.567859+00:00 app[web.1]: at TLSWrap.ssl.onhandshakedone (_tls_wrap.js:711:12) {
2021-01-19T08:53:42.567859+00:00 app[web.1]: code: 'DEPTH_ZERO_SELF_SIGNED_CERT'
2021-01-19T08:53:42.567860+00:00 app[web.1]: },
2021-01-19T08:53:42.567861+00:00 app[web.1]: original: Error: self signed certificate
2021-01-19T08:53:42.567861+00:00 app[web.1]: at TLSSocket.onConnectSecure (_tls_wrap.js:1502:34)
2021-01-19T08:53:42.567862+00:00 app[web.1]: at TLSSocket.emit (events.js:314:20)
2021-01-19T08:53:42.567862+00:00 app[web.1]: at TLSSocket._finishInit (_tls_wrap.js:937:8)
2021-01-19T08:53:42.567862+00:00 app[web.1]: at TLSWrap.ssl.onhandshakedone (_tls_wrap.js:711:12) {
2021-01-19T08:53:42.567863+00:00 app[web.1]: code: 'DEPTH_ZERO_SELF_SIGNED_CERT'
2021-01-19T08:53:42.567863+00:00 app[web.1]: }
2021-01-19T08:53:42.567864+00:00 app[web.1]: }
```

Here's my code. I've set `rejectUnauthorized: false` but that hasn't helped me. Please advise.

```
let sequelize = new (<any>Sequelize)(
    getConfig().DATABASE_DATABASE,
    getConfig().DATABASE_USERNAME,
    getConfig().DATABASE_PASSWORD,
    {
      host: getConfig().DATABASE_HOST,
      dialect: getConfig().DATABASE_DIALECT,
      port: 5432,
      dialectOptions: {
        ssl: true,
        rejectUnauthorized: false
      },
      logging:
        getConfig().DATABASE_LOGGING === 'true'
          ? (log) =>
              console.log(
                highlight(log, {
                  language: 'sql',
                  ignoreIllegals: true,
                }),
              )
          : false,
    },
  );
  ```

Thanks for everyones help in advance~! <3
  
  
<img width="1108" alt="Screen Shot 2021-01-19 at 3 53 59 PM" src="https://user-images.githubusercontent.com/16382165/105010962-f350e700-5a6e-11eb-80d8-9ebe88e76ca7.png">

**Comment by richardbiddle** [comment]
Thanks @brianc for your work on this. Using PGSSLMODE=no-verify helped but this lead to issues with psql:
```
$ export PGSSLMODE=no-verify
$ psql
psql: error: invalid sslmode value: "no-verify"
```
I am going to need to return to PGSSLMODE=require. This causes psql to allow self signed certs. 
https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-CONNECT-SSLMODE
`require: only try an SSL connection. If a root CA file is present, verify the certificate in the same way as if verify-ca was specified`

Maybe node-postgres could follow this same logic. If a root CA  (ssl: {ca: ) is present then {rejectUnauthorized: true} by default, otherwise {rejectUnauthorized: false}

**Comment by vtamm** [comment]
> Here's my code. I've set `rejectUnauthorized: false` but that hasn't helped me. Please advise.
> 
> ```
> let sequelize = new (<any>Sequelize)(
>     getConfig().DATABASE_DATABASE,
>     getConfig().DATABASE_USERNAME,
>     getConfig().DATABASE_PASSWORD,
>     {
>       host: getConfig().DATABASE_HOST,
>       dialect: getConfig().DATABASE_DIALECT,
>       port: 5432,
>       dialectOptions: {
>         ssl: true,
>         rejectUnauthorized: false
>       },
>       logging:
>         getConfig().DATABASE_LOGGING === 'true'
>           ? (log) =>
>               console.log(
>                 highlight(log, {
>                   language: 'sql',
>                   ignoreIllegals: true,
>                 }),
>               )
>           : false,
>     },
>   );
> ```
> 
> Thanks for everyones help in advance~! <3

@PrimeTimeTran I think you have an error in your `dialectOptions`. The `rejectUnauthorized` property should be located inside the `ssl` object, like so:
```
dialectOptions: {
  ssl: {
    require: true,
    rejectUnauthorized: false
}
```

This configuration solved both the `self-signed certificate` error as well as another error message (`psql: FATAL: no pg_hba.conf entry for host` I received when trying to run Sequelize migrations on a Postgres database (Heroku Postgres) attached to a Heroku dyno.

**Comment by cpedersen** [comment]
Hello, Is there any recent status on this issue? I have struggled for many days now with trying to get a DB connection working using node.js and Heroku -- using many different ssl settings:

require: true
rejectUnauthorized: false
ssl: true
pgsslmode: no-verify

Nothing is working. I get this error:

Error error: no pg_hba.conf entry for host "54.172.193.178", user "wjjjcrerskmfmg", database "dte2fdjjs0vdm", SSL off

I have Heroku Support trying to help me along with various developers that I work with. Nothing is working.

**Comment by charmander** [comment]
@cpedersen If you’re always getting the same error – specifically one that includes “SSL off” with the configuration `ssl: true` – you might be putting the settings in the wrong place. Can you show the context?

The complete code of a minimal app that reproduces the problem would also help.

**Comment by cpedersen** [comment]
I actually had ssl set to true in my package.json and had rejectAuthorized set to false in my postgrator-config.js file -- to no avail. What finally got me unblocked was setting 'PGSSLMODE' to 'no-verify' on the Heroku using their config vars. 

1. Add 'ssl=true' to package.json

"migrate:production": "env ssl=true DATABASE_URL=$(heroku config:get DATABASE_URL -a stormy-citadel-96294) npm run migrate",

2. Add the following to the module.exports part of the postgrator-config.js file --

  ssl: {
    rejectUnauthorized: false,
  },

3. Add the PGSSLMODE config var to the Heroku dyno, setting it to 'no-verify'.

Now I can connect to the Heroku DB from the Vercel client.

(Note that using '?PGSSLMODE=no-verify' as part of the DATABASE_URL in the code did not work for me.)

**Comment by Dzirt2006** [comment]
this topic helped me https://stackoverflow.com/questions/61097695/self-signed-certificate-error-during-query-the-heroku-hosted-postgres-database/66773313#66773313
my solution:
`const db = new Sequelize(
process.env.DATABASE_URL ||
`postgres://postgres:w2w2@localhost:5432/${databaseName}`, 
{
    logging: false,
    ssl: { rejectUnauthorized: false } //solved the problem with self signed sertificate
} `

**Comment by WithoutATowel** [comment]
Hi all! I'm running into the same issue, but none of the recommended fixes seem to work.

### My setup:

 - Node v15.12.0
 - Postgres database ("Heroku Postgres" add-on)
 - pg v8.5.1
 - sequelize v6.6.2
 - sequelize-cli v6.2.0 

### My Sequelize connection parameters:
```
{
  "development": {
    "username": "XXXXXX",
    "password": "XXXXXX",
    "database": "party_playlist",
    "host": "127.0.0.1",
    "dialect": "postgres"
  },
  "production": {
    "use_env_variable": "DATABASE_URL",
    "dialect": "postgres",
    "dialectOptions": {
      "ssl": {
         "require": "true",
         "rejectUnauthorized": "false"
      }
    }
  }
}
```

### What I've tried:
 1. Adding `rejectUnauthorized: false` to my Sequelize connection params
 2. Adding `ssl:true` to the config outside of dialectOptions
 3. Adding `NODE_TLS_REJECT_UNAUTHORIZED=0` to my Heroku environment variables
 4. Adding `PGSSLMODE=no-verify` to my Heroku environment variables
 5. Toggling Heroku's `Automatic Certificate Management` feature on/off 
 6. Rolling back to pg v7 (various subversions)

1-5 have no effect, and rolling back to pg v7 breaks the app (locally, any attempted reads/writes to the database hang with no error message; on heroku, the db:migrate command runs without an error message, but the db isn't updated). I assume that pg v7 is incompatible with my Node version or some other package version in my project, but I have no idea how to figure out what a compatible version set would be other than by trial and error, which isn't feasible. Also, as a side note, this is a hobby project so I'm not worried about MITM attacks.

Interestingly, I'm able to connect to my Heroku database just fine when I run `heroku pg:psql`!

Any ideas how I can fix this one? Any and all help would be much appreciated! Also [posted this to StackOverflow](https://stackoverflow.com/questions/66923586/self-signed-certificate-error-on-heroku-running-sequelize-dbmigrate) if folks would like internet points.

**Comment by charmander** [comment]
@WithoutATowel If it’s being passed directly to pg, the value of `"ssl"` should be `{ "rejectUnauthorized": false }`, not `{ "require": "true", "rejectUnauthorized": "false" }` (note the quotes).

**Comment by WithoutATowel** [comment]
Lol... Thank you @charmander!

**Comment by wamujlb** [comment]
@adrianhorning08 hi! I also use DO managed database but it doesn't work for me. I don't understand what I'm doing wrong.
Thank you in advance
```
export const pool = new Pool({
  user: DB_CONFIG.USER,
  password: DB_CONFIG.PASSWORD,
  host: DB_CONFIG.HOST,
  port: DB_CONFIG.PORT,
  database: DB_CONFIG.DATABASE,
  ssl: {
    rejectUnauthorized: true,
    ca: DB_CONFIG.CA_CERT,
  },
});
```

**Comment by sted6** [comment]
Hi all,

Having the same issue here. Even after setting dialectOptions as suggested previously in the thread.

I've got;

```
const proConfig = {
  use_env_variable: "DATABASE_URL",
  dialect: "postgres",
  dialectOptions: {
    ssl: {
      require: true,
      rejectUnautorized: false,
    },
  },
};

const db =
  process.env.NODE_ENV === "production"
    ? new Sequelize(process.env.DATABASE_URL, proConfig)
    : new Sequelize(
        process.env.DB_NAME,
        process.env.DB_USER,
        process.env.DP_PASSWORD,
        {
          dialect: "postgres",
          host: process.env.DB_HOST,
          port: process.env.DB_PORT,
        }
      );
```
And I'm still getting: 

> UnhandledPromiseRejectionWarning: SequelizeConnectionError: self signed certificate

Is there anything else I could try?

**Comment by charmander** [comment]
@sted6 The spelling is `rejectUnauthorized` with an `h`. Also, `require: true` isn’t a thing, so you can remove that.

**Comment by sted6** [comment]
@charmander uhg I could of swore I even copy pasted that. Thank you so much.

**Comment by forrestpark** [comment]
Hi, I've tried all mentioned options but am still having issues. I've posted my issue here: https://stackoverflow.com/questions/68327934/error-self-signed-certificate-while-running-sequelize-dbmigrate-on-heroku. Any help would be greatly appreciated. Thank you!

**Comment by charmander** [comment]
> 2021-07-10T07:06:18.758245+00:00 app[web.1]: dialectOptions: { ssl: { rejectUnautorized: false } }

> ```js
> const production = {
>   "use_env_variable": "DATABASE_URL",
>   "dialect": "postgres",
>   "dialectOptions": {
>     "ssl": {
>       "rejectUnautorized": false,
>     }
>   }
> }
> ```

@forrestpark Same typo as @sted6, it’s `rejectUnauthorized` with an `h`.

**Comment by simon1tan** [comment]
The real solution is to use prisma https://github.com/prisma/prisma

**Comment by RodrigoBeiraoCardoso** [comment]
hello everyone, i have the same problem and ive tried litterally everything i found, from global variables on heroku to all kinds of those params: const sequelize = new Sequelize(database, user, password, {
  host,
  port,
  dialect: 'postgres',
  dialectOptions: {
    ssl: {
      rejectUnauthorized: "false" 
    },
  },
  logging: false
})
ive tried it that way  like everyone else, ive also had ssl:true, and the require:true and that part outside the dialect options just the ssl {reject..} ive also tried reverting pg to a previous version, ahh ive put on heroku 2 global variables one USE_SSL=true and another one that was noded_reject_etc= 0 (cant remenber the right name) and nothing works nothing ahah
I really need help guys would apreciate anyone who can help

**Comment by jsanta** [comment]
@RodrigoBeiraoCardoso

> rejectUnauthorized: "false"

Should be `rejectUnauthorized: false ` without quotes. A boolean is needed not a string ;)
Everything else looks fine to me.

**Comment by RodrigoBeiraoCardoso** [comment]
> @RodrigoBeiraoCardoso
> 
> > rejectUnauthorized: "false"
> 
> Should be `rejectUnauthorized: false ` without quotes. A boolean is needed not a string ;)
> Everything else looks fine to me.

yeahh that was just a mistake that i copy pasted ahah ive tried without it and still the same error, ive tried like this dialectOptions: {
    ssl: {
      rejectUnauthorized: false // <<<<<<< YOU NEED THIS
    }
  },
  ssl: true,
Like this dialectOptions: {
    ssl: {
      require: true
      rejectUnauthorized: false // <<<<<<< YOU NEED THIS
    }
  },
  ssl: true,
and like all that but without the ssl:true
also tried just the  ssl: {
      rejectUnauthorized: false // <<<<<<< YOU NEED THIS
    },
withouth being inside the dialect and nothing works.
The grobal variables ive tried on heroku are USE_SSL =true (which i still use) and NODE_TLS_REJECT_UNAUTHORIZED = 0 (wich im not using anymor)

**Comment by Tosinkoa** [comment]
Pls help, I have try all the answer here, nothing seems to work for me.
I've spent 2days on this with no avails, pls help.
After adding all answers here, this is was what I tried as a last attempt.
```
import pkg from 'pg';
const { Pool } = pkg;

const pool = new Pool({
  user: process.env.USER,
  host: process.env.HOST,
  database: process.env.DATABASE,
  password: process.env.PASSWORD,
  port: process.env.PORT,
  ssl: {
    require: true,
    rejectUnauthorized: false
  }
})

console.log("Connected to database... ✅✅✅")
export default pool
```
```
import express from "express"
import rootRoute from "./src/root_Route.js"
import cookieParser from "cookie-parser"
import passport from "passport"
import connectPgSimple from "connect-pg-simple"
import session from "express-session"
import dotenv from "dotenv"
import cors from "cors"
import "./src/LIB/DB-Client.js"
import "./src/PASSPORT_STRATEGY/google-auth-strategy.js"
import "./src/PASSPORT_STRATEGY/facebook-auth-strategy.js"
import pool from "./src/LIB/DB-Client.js"
dotenv.config()
const app = express()

app.use(
  cors({
    origin: ["http://localhost:3000", "https://frontend.netlify.app"],
    credentials: true,
    methods: "GET, PUT, POST, DELETE",
    optionsSuccessStatus: 200
  })
)
const PgStore = connectPgSimple(session)
const store = new PgStore({ conString: process.env.DATABASE_URL, schemaName: "hidden", createTableIfMissing: true })

app.use(express.json())
app.use(cookieParser())
app.set("trust proxy", 1)

app.use(
  session({
    store: store,
    secret: "myscecret",
    saveUninitialized: false,
    resave: false,
    cookie: {
      maxAge: 1000 * 60 * 60 * 24,
      secure: true,
      httpOnly: true,
      sameSite: "none",
    },
  })
)

app.get("/", (req, res) => {
  const userId = req.session.user ?? req.user
  res.send("API Running...")
})

app.use(passport.initialize())
app.use(passport.session())
rootRoute(app)

const PORT = process.env.PORT || 4000

app.listen(PORT, (req, res) => console.log(`Server running on PORT:${PORT}...`))
```
I also added this as an env in heroku `PGSSLMODE=no-verify`, But still I get this error: `Failed to prune sessions: self signed certificate`

Thanks in advance 🙏🙏

**Comment by charmander** [comment]
@Tosinkoa You haven’t passed the pool to connect-pg-simple in that code, so none of the configuration other than `DATABASE_URL` is being used.

**Comment by Tosinkoa** [comment]
Everything works fine now, thank you so much.

> @Tosinkoa You haven’t passed the pool to connect-pg-simple in that code, so none of the configuration other than `DATABASE_URL` is being used.

**Comment by pantelis-karamolegkos** [comment]
I have the following `node` script

```js

const { Pool } = require('pg');
const { parse } = require('pg-connection-string');

const config = parse(process.env.DATABASE_URL);
config.ssl = { rejectUnauthorized: true };

const pool = new Pool(config);

pool.connect((err, client, done) => {
  if (err) {
    console.error('Error acquiring client', err.stack);
    return;
  }

  const testQuery = 'SELECT * FROM information_schema.tables';
  client.query(testQuery, (err, result) => {
    done();
    if (err) {
      console.error('Error executing query', err.stack);
      return;
    }
    console.log('Query result:', result.rows);
    pool.end();
  });
});
```

Before running the script I am exporting the `NODE_EXTRA_CA_CERTS=../aws-global-bundle.pem` where the `.pem` file is [this][1] one.

My question is how can I enforce the `verify-full` option as described in the official `postgresql` [documentation][2] and what is the practical difference with setting the `rejectUnauthorized` to `true` ?


I have tried the following

```js
config.ssl = { 
  rejectUnauthorized: true,
  sslmode: 'verify-full',
  ca: fs.readFileSync('./global-bundle.pem').toString(),
};
```

However it is as if the `sslmode: 'verify-full',` is not actually being taken into account because when I use `sslmode: 'foobar',` it does not complain at all (and connects as expected.

  [1]: https://truststore.pki.rds.amazonaws.com/global/global-bundle.pem
  [2]: https://www.postgresql.org/docs/current/libpq-ssl.html#LIBQ-SSL-CERTIFICATES

**Comment by charmander** [comment]
@pantelis-karamolegkos `sslmode` is a connection string parameter (and has a corresponding environment variable named `PGSSLMODE`). It’s translated to the `ssl: …` property of the pg configuration object when you use a connection string. There is no `ssl.sslmode`. `rejectUnauthorized: true`/verify-full is the default behavior of [Node’s tls module](https://nodejs.org/api/tls.html#tlsconnectoptions-callback) that pg passes the `ssl` object to directly, so all you need is

```js
config.ssl = {
  ca: fs.readFileSync('./global-bundle.pem'),
};
```
